### PR TITLE
Fix `src/main/kotlin` not being recognised by default

### LINKED ISSUE #228: Doesn't work in `src/main/kotlin`
After AGP 7.0 we can use `src/main/kotlin` as kotlin sourceSet by default, but the kotlinter doesn't work in this location.

See JLLeitschuh/ktlint-gradle#524.

### PR DESCRIPTION
**Pull request by mateuszkwiecinski**
Fixes #228

Seeing the linked issue, I was curious if that's really the case 👀 
I wanted to reproduce the issue in tests, which led me to bumping things to make the whole project be compiled against AGP 7.X and to basically make the project up-to-date again. 
Not sure if you like the changes I'm proposing here. I tried to keep them in separate commits, just in case you want to take only some of them. 

Important note, the changes requires consumers to use AGP 7.0 and up, due to usage of the `.kotlin` extension. If the goal is to maintain old Gradle compatibility then I guess it's required to access kotlin sourcesets in a smarter way)

### CODE REVIEW

File: .github/workflows/default.yml
Comment (mateuszkwiecinski) [inline .github/workflows/default.yml:53]: I thought it makes sense to test on Java 17. I expect more people start using it as it is an LTS version

File: build.gradle.kts
Comment (mateuszkwiecinski) [inline build.gradle.kts:37]: `compileOnly` isn't resolvable anymore. I explicitly declare new configuration which can be used via `pluginUnderTestMetadata` task

File: build.gradle.kts
Comment (mateuszkwiecinski) [inline build.gradle.kts:70]: I didn't have Java 8  installed, so I thought it makes sense to set recent LTS to be built with, but still target Java 8 (the `options.release` call below)

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt
Comment (mateuszkwiecinski) [inline src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt:26]: `sourceSet.kotlin` is accessible only on AGP 7.0 and up. AFAIU this wil fail if consumer's project uses lower version. This line is also a fix for the linked issue.

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/KotlinJvmSourceSetApplier.kt
Comment (mateuszkwiecinski) [inline src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/KotlinJvmSourceSetApplier.kt:14]: As a boy-scout rule, I replaced non-lazy apis usages with lazy ones

File: src/test/kotlin/org/jmailen/gradle/kotlinter/functional/AndroidProjectTest.kt
Comment (mateuszkwiecinski) [inline src/test/kotlin/org/jmailen/gradle/kotlinter/functional/AndroidProjectTest.kt:79]: If needed, one can also add test case fo `src/xxx/java`, to verify `/java` directory is also recognised properly

File: src/test/kotlin/org/jmailen/gradle/kotlinter/functional/CustomTaskTest.kt
Comment (mateuszkwiecinski) [inline src/test/kotlin/org/jmailen/gradle/kotlinter/functional/CustomTaskTest.kt:24]: I also noticed that I unnecessairly added a dependency, which can be easily replaced with comments

File: build.gradle.kts
Comment (mateuszkwiecinski) [inline build.gradle.kts:45]: Ah, forgot to mention that one. That's basically the same as in #201, with the same reasoning, but this time _something_ in tests config uses incompatible version of kotlin, so instead of digging which dependency needs to be bumped I'm propsoing to simply keep them consistent 🤷‍♂️

File: .github/workflows/default.yml
Comment (jeremymailen) [reply .github/workflows/default.yml:53]: Agree

File: build.gradle.kts
Comment (jeremymailen) [reply build.gradle.kts:70]: Sounds good

File: .github/workflows/default.yml
Comment (jeremymailen) [inline .github/workflows/default.yml:53]: Do we also need to update the release.yml workflow to use Java 17?

File: build.gradle.kts
Comment (jeremymailen) [inline build.gradle.kts:30]: Yeah, it's probably good to update. I hadn't yet because afaik nothing in 0.43.1 and 0.43.2 impacted kotlinter.
Still waiting on a fix to wildcard imports :/.

File: build.gradle.kts
Comment (jeremymailen) [inline build.gradle.kts:28]: Yeah, there's probably no way around losing compatibility with prior versions of AGP since we use the plugin itself to resolve the source sets and can't easily include two in the compile.

Can you go ahead and add an Android Plugin column to the compatibility matrix?
https://github.com/jeremymailen/kotlinter-gradle#compatibility

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/KotlinJvmSourceSetApplier.kt
Comment (jeremymailen) [reply src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/KotlinJvmSourceSetApplier.kt:14]: Yes, in fact we had needed to be lazy in a few places in the code because if someone customizes the source sets in their build file we need to resolve our targets after that is configured, otherwise in early versions of kotlinter we didn't pick up the customization. I assume this is harmonious with that?

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt
Comment (jeremymailen) [reply src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt:26]: Yes, I guess we'll need a compatibility boundary as per my earlier comment.

**Review by jeremymailen — COMMENTED** [review]
Thank you as always for your diligence in watching over the project!

This looks good to go with the Readme update and release workflow bump. Thank you!

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/KotlinJvmSourceSetApplier.kt
Comment (mateuszkwiecinski) [reply src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/KotlinJvmSourceSetApplier.kt:14]: AFAIR The old issue was about customization of an existing sourceset not being picked up. This change will "fix" the plugin if someone _registers new_ sourceset after kotlinter plugin gets applied. So yeah, it's definitelly consistent with the other behavior 👍

File: .github/workflows/default.yml
Comment (mateuszkwiecinski) [reply .github/workflows/default.yml:53]: Good catch 👍 I clearly missed that!

File: build.gradle.kts
Comment (mateuszkwiecinski) [reply build.gradle.kts:28]: I pushed another commit, with a "fix" which would maintain old behavior for AGP's `4.x` users. I also added `integration-tests-android` job to existing workflow which makes sure the plugin will work with older AGP versions.

image

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt
Comment (mateuszkwiecinski) [inline src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt:28]: This is a _trick_ which I read as "if there is a speparate kotlin sourceset append it to java ones". I initially wanted to do a version check on AGP, but I couldn't find a stable api to do so. 
I once reported: https://issuetracker.google.com/issues/188859855 where I tried to let AGP maintainers know it's unclear what's the preferred way to obtain AGP version. 
The alternatives I considered are:
- add their internal artifact and use Version object
- continue using deprecated property until they remove it at some point
- do a reflective call to check if `kotlin` property exists

As a conclusion I'm proposing to do catch and ignore the exception if AGP 7.0 specific classes aren't available in runtime

File: build.gradle.kts
Comment (jeremymailen) [reply build.gradle.kts:28]: Very nice

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt
Comment (jeremymailen) [reply src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt:28]: That seems like a reasonable approach all considered. This or that.

File: .github/workflows/default.yml
Comment (jeremymailen) [inline .github/workflows/default.yml:110]:
```suggestion
          build-root-directory: test-project-android
```

File: .github/workflows/default.yml
Comment (jeremymailen) [inline .github/workflows/default.yml:94]: This version isn't found. Perhaps it requires a different repo or a different patch version?
I'm fine dropping that from the matrix if it end up being a pain.

File: src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt
Comment (jeremymailen) [inline src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt:26]: Let's put a comment above this line noting why. Something like "detect kotlin source paths supported by AGP 7 and later"

**Review by jeremymailen — COMMENTED** [review]
NIcely done. A couple of fixes.
Once we land this I'll probably push a minor release.

File: .github/workflows/default.yml
Comment (mateuszkwiecinski) [reply .github/workflows/default.yml:94]: Yeah, I wasn't sure what's the lower boundary so I picked a version I consider absolute minimum and I would be surprised if any continously maintained project used lower version. I know some project refrain from migrating to 7.0 because it still has some hickups and I _heard_ the 7.2 should be considered the "proper" version of AGP 7 (7.2 is still in alpha and is planned for mid February) 

My initial setup didn't work because older AGP versions don't publish gradle metadata needed for `plugins` block ([Issuetracker](https://issuetracker.google.com/issues/64551265)). I dropped `plugins`  block in favor of custom `resolutionStrategy`[here](https://github.com/jeremymailen/kotlinter-gradle/pull/229/files#diff-124a21ab4c45e752a2d3151e279332a337c772e9198fdd248dd43a1127c3e6c9R8)

>I'm fine dropping that from the matrix if it end up being a pain.

It wasn't that big pain so I decided to include necessary changes. My motivation was to explicitly support 4.0 to see the workflow failing when some changes in Kotlinter plugin will introduce breaking changes, so it can be mentioned in the release notes

**Review by jeremymailen — APPROVED** [review]

**Comment by JLLeitschuh** [comment]
PRs are welcome to resolve this issue, please do include test coverage.

**Pull request merged**
